versiondb/client/changeset.go

```diff
@@ -0,0 +1,322 @@
+//
+// layout: deletion(1) + keyLen(2) + key + [ valueLen(4) + value ]
+func encodedSizeOfKVPair(pair iavl.KVPair) int {
+	size := 3 + len(pair.Key)
+	if !pair.Delete {
+		size += 4 + len(pair.Value)
+	}
+	return size
+}
+
+// encodeKVPair encode a key-value pair in change set.
+// see godoc of `encodedSizeOfKVPair` for layout description,
+// returns error if key/value length overflows.
+func encodeKVPair(pair iavl.KVPair) ([]byte, error) {
+	buf := make([]byte, encodedSizeOfKVPair(pair))
+
+	if len(pair.Key) > math.MaxUint16 {
+		return nil, fmt.Errorf("key length %d is too big", len(pair.Key))
+	}
+
+	binary.LittleEndian.PutUint16(buf[1:], uint16(len(pair.Key)))
+	copy(buf[3:], pair.Key)
+	if pair.Delete {
+		buf[0] = 1
+	} else {
+		if len(pair.Value) > math.MaxUint32 {
+			return nil, fmt.Errorf("value length %d is too big", len(pair.Value))
+		}
+		offset := 3 + len(pair.Key)
+		binary.LittleEndian.PutUint32(buf[offset:], uint32(len(pair.Value)))
+		copy(buf[offset+4:], pair.Value)
+	}
+
+	return buf, nil
+}
+
+// readKVPair decode a key-value pair from reader
+//
+// see godoc of `encodedSizeOfKVPair` for layout description
+func readKVPair(reader io.Reader) (*iavl.KVPair, error) {
+	var buf [4]byte
+	if _, err := io.ReadFull(reader, buf[:3]); err != nil {
+		return nil, err
+	}
+
+	var pair iavl.KVPair
+	pair.Delete = (buf[0] == 1)
+
+	keyLen := binary.LittleEndian.Uint16(buf[1:])
+	pair.Key = make([]byte, keyLen)
+	if _, err := io.ReadFull(reader, pair.Key); err != nil {
+		return nil, err
+	}
+
+	if !pair.Delete {
+		if _, err := io.ReadFull(reader, buf[:]); err != nil {
+			return nil, err
+		}
+		valueLen := binary.LittleEndian.Uint32(buf[:])
+
+		pair.Value = make([]byte, valueLen)
+		if _, err := io.ReadFull(reader, pair.Value); err != nil {
+			return nil, err
+		}
+	}
+
+	return &pair, nil
+}
+
+// openChangeSetFile opens change set file,
+// it handles compressed files automatically,
+// also supports special name "-" to specify stdin.
+func openChangeSetFile(fileName string) (io.ReadCloser, error) {
+	if fileName == "-" {
+		return os.Stdin, nil
+	}
+	var reader io.Reader
+	fp, err := os.Open(fileName)
+	if err != nil {
+		return nil, err
+	}
+	switch {
+	case strings.HasSuffix(fileName, ZlibFileSuffix):
+		zreader, err := zlib.NewReader(fp)
+		if err != nil {
+			_ = fp.Close()
+			return nil, err
+		}
+		reader = bufio.NewReader(zreader)
+	case strings.HasSuffix(fileName, SnappyFileSuffix):
+		reader = snappy.NewReader(fp)
+	default:
+		reader = bufio.NewReader(fp)
+	}
+	return WrapReader(reader, fp), nil
+}
+
+// withChangeSetFile opens change set file and pass the reader to callback,
+// it closes the file immediatelly after callback returns.
+func withChangeSetFile(fileName string, fn func(io.Reader) error) error {
+	reader, err := openChangeSetFile(fileName)
+	if err != nil {
+		return err
+	}
+	defer reader.Close()
+	return fn(reader)
+}
+
+// IterateChangeSets iterate the change set files,
+func IterateChangeSets(
```

Comment:
this function may take a long time to process, right? Should we add `context` if the user wants to interrupt the process?